### DISCUSSION TITLE
Which Ollama models are supported?

### DISCUSSION THREAD

**Discussion by mihow** [cline/cline#241] [Feature Requests]
Thank you so much for adding support for local Ollama models!

I was hoping to use `deepseek-coder-v2` or `command-r`, but I get this error when trying either:

"400 command-r:latest does not support tools"

`llama3.1` is able to generate some code, but at the end, or in the middle of the task usually errors with something like:

> Since you haven't provided any specific problem or scenario for me to solve, I'll demonstrate how to use the attempt_completion tool based on a hypothetical situation.
> 
> Let's say we've completed a task and want to present the result to the user. In this case, we can use the attempt_completion tool to show the final result.
> ...

> Claude is having trouble...
> This may indicate a failure in his thought process or inability to use a tool properly, which can be mitigated with some user guidance (e.g. "let's try breaking this large file down into smaller files").

**Comment by nomadictuba2005** [comment]
ya same thing with me?

**Comment by psalzman** [comment]
I haven't found one that can execute the commands however Qwen2.5-coder seems to function conversationally.

**Reply by TomLucidor** [reply]
The short context length issue is seen in OpenHands  (inability to read the whole instruction prompt + context code), maybe the same applies to Cline?

**Comment by hhao** [comment]
ollama run [hhao/qwen2.5-coder-tools](https://ollama.com/hhao/qwen2.5-coder-tools)

you can run this model  with Claude Dev, it's free.

![image](https://github.com/user-attachments/assets/c9e24db3-21de-44a2-bdae-7135b4bb4fe8)

**Reply by psalzman** [reply]
Nice. I see it was updated about a minute ago.  I'll pull it down and play a bit.

**Reply by ChristianWeyer** [reply]
> Nice. I see it was updated about a minute ago. I'll pull it down and play a bit.

Did it do what you wanted it to do? :-)

**Reply by hotwa** [reply]
I'll pull it down and play a bit, too

**Reply by daaku** [reply]
I gave it a quick shot and it worked - I was wondering if you could share how you made this model? I'd like to make a `Q4_K_M` quantized one because my laptop is not beefy enough for the `Q8` :)

**Reply by damarges** [reply]
edit_: got it sorted out. switching the model while in a task does not work. create a new task with the new model selected made it work.

edit2: doesn't work. cline always claims that the prompt it uses are to complex. it loops itself with that message. 

Well doesn't work for me with the model:

![image](https://github.com/user-attachments/assets/bfa29b0e-187f-4f22-87e7-5b80a031455e)

**Reply by phalexo** [reply]
> ollama run [hhao/qwen2.5-coder-tools](https://ollama.com/hhao/qwen2.5-coder-tools)
> 
> you can run this model with Claude Dev, it's free.
> 
>
I have set up the template and system prompt to use with Qwen2.5-Coder-32b-Instruct.Q8_0.gguf 
and it mostly works pretty well.

It has a problem with "list_code_definition_names" specifically with Javascript code, Express/Node, because seemingly either an LLM or
Cline do not understand the following format:

exports.saveEmployee = async (req, res) => { /* stuff */ }

it does understand

async function saveCandidate(req, res) { /* stuff */ }

module.exports = {
    saveCandidate,
    otherFunc
};

or 

exports.saveCandidate = saveCandidate;
exports.otheFunc = otherFunc;

The problem is that it generates code, that its own TOOL does not understand, although still valid code.

I don't know if this affects other languages too.

Maybe there is a way to tweak the system prompt/template for Qwen models to better understand different declaration formats, @hhao

**Reply by wpp-bassam-assoum** [reply]
@hhao I tried using [hhao/qwen2.5-coder-tools](https://ollama.com/hhao/qwen2.5-coder-tools) in Ollama with Cline, but I am getting this error :
`400 json: cannot unmarshal array into Go struct field Message.messages.content of type string`

These are my Cline settings:
<img width="511" alt="image" src="https://github.com/user-attachments/assets/92cd9678-29ed-4ffa-a8d5-4d67f6035d65">\

Any idea what should I do for it to work? I even tried creating images from Modelfile using 32k context

EDIT: I'm using a Mac M2 Pro machine with 32GB ram

**Comment by erkkimon** [comment]
+1

**Reply by brightleo** [reply]
+1
It's not possible to execute with local ollama.（Windows）

**Comment by isaacveg** [comment]
It can. What you should do is increase the context length of your ollama model. The prompt of Cline is VERY LONG and 32768 is not enough to read in all the system prompt and your prompt.

Do the following steps:
1. create a model file, i.e. new_model_file, with the following:
```
# Modelfile
FROM qwen2.5:14b
PARAMETER num_ctx 65536
```
change the model name after FROM to your model, and the context length after num_ctx. On my 36gb m3pro MacBook, the context length is reasonable and 131072 is too much for my computer. If you want to switch back, adjust the num_ctx to your original length. Use ollama show MODELNAME to check. More information at !(https://github.com/ollama/ollama/blob/main/docs/modelfile.md)

2. apply the model file. ```ollama create -f new_model_file qwen2.5:14b```

3. Now you can use Cline.

**Reply by phalexo** [reply]
> It can. What you should do is increase the context length of your ollama model. The prompt of Cline is VERY LONG and 32768 is not enough to read in all the system prompt and your prompt.
> 
> Do the following steps:
> 
>     1. create a model file, i.e. new_model_file, with the following:
> 
> 
> ```
> # Modelfile
> FROM qwen2.5:14b
> PARAMETER num_ctx 65536
> ```
> 
> change the model name after FROM to your model, and the context length after num_ctx. On my 36gb m3pro MacBook, the context length is reasonable and 131072 is too much for my computer. If you want to switch back, adjust the num_ctx to your original length. Use ollama show MODELNAME to check. More information at !(https://github.com/ollama/ollama/blob/main/docs/modelfile.md)
> 
>     2. apply the model file. `ollama create -f new_model_file qwen2.5:14b`
> 
>     3. Now you can use Cline.

If you had a choice between 14B model/64K context vs. 32B model/20K context?  I have only 49GB VRAM spread over 4 GPUs.

**Reply by isaacveg** [reply]
> If you had a choice between 14B model/64K context vs. 32B model/20K context? I have only 49GB VRAM spread over 4 GPUs.

For me I prefer using 4bit quant and 14b model is still too big for 27gb usable vram. I suggest using a smaller model with large context window, since the kvcache you’re using will take more and more space as your conversation going. In your situation, I would choose 14B and 64 or even longer context.

**Reply by bardackx** [reply]
This worked! The vscode plugin should help you do this

**Comment by phalexo** [comment]
I copied the template, system prompt, but I had to play with parameter
stop, ctx_size, gpu_num to fit my hardware. I am using Q8_0 version.

**Comment by phalexo** [comment]
Actually I also remember having to tweak a little a line in the system
prompt.

**Comment by brightleo** [comment]
[ollama run hhao/qwen2.5-coder-tools]  is too slow, 
[ollama run maryasov/qwen2.5-coder-cline] has some issues but runs smoothly.

**Comment by phalexo** [comment]
It is kind of slow, but my main issue is that the model itself has
problems. For example, it deletes sections of code that were working just
fine, and are not even related to my prompt/request. Not much time saving
if I have to constantly monitor everything myself to prevent it from
causing damage.

**Comment by erkkimon** [comment]
Just giving my five cents here: There are cline-specific models in ollama website and you can search with phare "cline" to find [them](https://ollama.com/search?q=cline). I tried [maryasov/qwen2.5-coder-cline](https://ollama.com/maryasov/qwen2.5-coder-cline) and it worked. Haven't tried them all yet, but please share your experiences in order to find the best one(s) collaboratively!

**Comment by phalexo** [comment]
How many BITS is the maryasov flavor? I am using the more general Q8_0
version, I think direct from Qwen.

**Comment by acidtib** [comment]
im getting pretty good results with 3b and 7b

https://ollama.com/acidtib/qwen2.5-coder-cline

**Comment by leporel** [comment]
I think if cline uses OpenAI compatible API, then they cant [set ctx size](https://github.com/ollama/ollama/blob/main/docs/openai.md#setting-the-context-size), that's why you need crate custom mod file
but if cline will use direct api for ollama, it possible to use any default model with large context from ollama

**Comment by hrqiang** [comment]
I ran in the same 400 error connecting to ollima.
```
400 json: cannot unmarshal array into Go struct field Message.messages.content of type string
```
Tried  different models and `qwen2.5:14b` with different context size, same error. 
Any instruction how to setup ollima with cline?

**Reply by acidtib** [reply]
give this model a try, ive had great success with it

https://ollama.com/acidtib/qwen2.5-coder-cline

**Reply by anhalu** [reply]
@acidtib  I see you change num_ctx to 16k and the system prompt, Also did you change anything? fine tune or something? have you tried with larger models? Thanks a lot.

**Comment by phalexo** [comment]
Is the Cline version better than the original Qwen 2.5 coder? I am using
the original model at q8_0.

**Reply by leporel** [reply]
they are same, but cline version have modfile with "num_ctx": 16384,

**Reply by acidtib** [reply]
qwen2.5-coder-cline is Qwen 2.5 coder it just has an updated system prompt and some param changes to make it usable with cline, i have only added 7b

**Comment by phalexo** [comment]
Ok, I set my context to 20000, but the question is more about different
quantizations. I am using 8 bits and Ollama uses 4 bits. I am wondering if
there is a difference in quality and if the 4 bit version was fine-tuned
somehow.

**Comment by elemental-mind** [comment]
This does not seem to be a feature request - can this be transferred into an issue @mihow ?

**Comment by anhalu** [comment]
Hey guys, has anyone tried qwq-32b-preview ?

**Comment by mihow** [comment]
Thanks for asking @elemental-mind. It probably should be an issue! The "feature requested" is information about what models are supported - or what specs does a model need to be supported. Right now the docs say "You can also configure any OpenAI compatible API, or use a local model through LM Studio/Ollama", which is misleading and requires many hours & gigabytes of troubleshooting. But I am guessing the maintainers do not know either, so hopefully this community discussion is helpful for anyone looking.

I just opened a PR with a few suggestions: https://github.com/cline/cline/pull/1084

**Comment by bulatoff555** [comment]
ollama run rashakol/sky-t1-32B-preview-cline. Works pretty good. But still 32K context. Why anyone don't think about creating middle ware to connect ollama nomic text embed (tokenisation method) and convert in real time chat history to some chunks maybe, or a configuration of vector db, maybe, ALSO,   ollama run hhao/qwen2.5-coder-tools:0.5b it is more than enough for creating evev full stack apps, in anything llm it gives pretty working code snippets, soooo let's convert its context length to 1 million, and give tools to it , be interact with cline. Excuse for my English.

**Comment by RafTurek** [comment]
Hello everyone.
I have checked few models in Vs code
Llama 3.2
DdepseekR1 -1,5b and 7b 
Normal qwen2.5-coder-cline:3b
None of them were able to properly make any code or suggestions 
Usually they're falling in some kind of loop writing all the time the same message
Later I've read here that maybe qwen2.5-coder-cline is specially made for this extension so I've tried it also 
7b and 3b
3b is ok working well started to make files on his own 
7b on my machine little to slow unfortunately.
But still they are falling in some kind of loop 
All the time busy with the same files and the same message and Sam code in the files.
Only I have this issue?
Any solutions?
Thanks 
Have a nice one!
R.

**Reply by narmaku** [reply]
I have also tried a couple of "-cline" ollama models and I have exactly the same issue. They succeed in the creation of files but then they regenerate the same file over and over again, and never stop on its own. I have no clue on how to solve this. I guess it can be related to the stop command that one can send to the model, but not entirely sure.

**Reply by RafTurek** [reply]
So probably only Claude works good with Cline or maybe got it to do with the hardware maybe?

**Reply by roycamp** [reply]
I also have the same issue where it will start okay and then begin outputting the same chunk of text over and over again.

**Reply by willf-sirge** [reply]
I have only been able to use cline in a functional coding capacity when using main stream api bases models. Like Gemini, deepseek, Open AI. Like many of you I am desperate for a free and local based alternative. I have tried all of ollama based models mentioned in this ticket and none of them come even close to being usable or matching the speed and accuracy of these other models. Trying to use ollama with cline in my experience is basically unusable. This honestly really sucks as I feel like this would probably be a major use case for this tool for the majority of the people out there. I don't know how many people can continue to burn through massive amounts of money through the commercial APIs

**Reply by erkkimon** [reply]
Agree. You might wanna try this one: https://ollama.com/tom_himanen/deepseek-r1-roo-cline-tools:32b

There are different model sizes available. If you have enough VRAM, it should work. Especially large tasks require lots of VRAM cause problems but keeping the tasks small enough, consumer hardware should be sufficient. I also choose smaller or bigger model size depending on the task – if the task is simple I use smaller model to preserve headspace for longer conversation with Cline, and if the task is cognitively more demanding, I try to keep the task very strictly scoped and use larger model.

I use RTX 3090 Ti but based on Deepseek R1 I'm convinced that it is worth it to purchase more VRAM for local LLM-programming.

**Comment by kdeugi** [comment]
i have follow this https://ollama.com/acidtib/qwen2.5-coder-cline so i can run this on local Window PC However i am getting  Error 405: Method Not Allowed please someone help on this?

**Comment by lemassykoi** [comment]
I'm using [Qwen3:30b](https://ollama.com/library/qwen3:30b) with context window set to 32768 and I have good results, with 12 Gigas VRAM

**Reply by BradKML** [reply]
Seconding this, but what about DeepSeek distills, are they good as well?

**Reply by andreybond** [reply]
How can you fit a 30b model into 12 GB VRAM? It should be roughly ~19-20Gb in q_4 (int4) mode..

**Reply by lemassykoi** [reply]
Ollama is offloading on GPU VRAM, then on RAM
![image](https://github.com/user-attachments/assets/8f7f389d-cdd3-41f1-80d3-c81bc1f14c5a)

**Reply by andreybond** [reply]
Wow! Is it default behavior or needs to be configured somehow? Never seen this... For me it is always CPU only if model does not fit GPU VRAM.

**Reply by lemassykoi** [reply]
> Wow! Is it default behavior or needs to be configured somehow? Never seen this... For me it is always CPU only if model does not fit GPU VRAM.

yes, default behavior for Ollama

**Comment by BradKML** [comment]
There are many models here by "tom_himanen", "maryasov", and "mychen76" so please test them accordingly https://ollama.com/search?q=RooCode https://ollama.com/search?q=cline
